Scaffolding for realtime-to-realtime

j0sh opened the pull request:
**What does this pull request do? Explain your changes. (required)**

Initial bits of the "live-video-to-video" pipeline (better name welcome)

Depends on https://github.com/livepeer/ai-worker/pull/229 for the realtime schema

Adds a MediaMTX handler for the gateway

Adds realtime-to-realtime capability to the orchestrator

Adds selection support between G/O

Orchestrator will respond to the job with trickle endpoints (publish, subscribe)

j0sh committed:
realtime: Add MediaMTX handler to G, and O caps and signaling

leszko commented on cmd/livepeer/starter/starter.go:
```diff
@@ -1348,6 +1348,15 @@ func StartLivepeer(ctx context.Context, cfg LivepeerConfig) {
 					if *cfg.Network != "offchain" {
 						n.SetBasePriceForCap("default", core.Capability_SegmentAnything2, config.ModelID, autoPrice)
 					}
+				case "realtime-to-realtime":
```
How about calling it `realtime-ai` or `realtime-inference` or `realtime-video-ai`?

j0sh replied:
Naming is the hardest thing in computer science :)

For reference here are the names of the other pipelines, which generally follow `<input-type>To<output-type>`

https://github.com/livepeer/go-livepeer/blob/439057979b22c6e6c4d6d388c2a811ee13fbcda3/core/capabilities.go#L75-L81

> How about calling it realtime-ai or realtime-inference or realtime-video-ai?

To me, the `-ai` or `-inference` suffixes don't fit any of the existing patterns and don't convey information in this context since all other pipelines are AI inference. We may also have other types of 'live' pipelines, so I think we want to specify the output type here too.

Another thing I have been thinking - we have been using the `realtime` terminology but I think `live` fits a little better (and is more on-brand for Livepeer). Plus it is slightly shorter.

So one option might be to use a prefix, eg `live-` for `live-video-to-video`

This makes the naming easily extendable to other areas too, eg `live-audio-to-text` for live transcription, `live-segment-anything-2` etc

rickstaa replied:
@j0sh I like the live prefixed input output naming scheme.

j0sh replied:
Renamed to `live-video-to-video` in https://github.com/livepeer/go-livepeer/pull/3210/commits/a92dacb663aa2f12a8446c457f8342877c3aa42e

leszko commented on server/ai_mediaserver.go:
```diff
@@ -72,6 +72,7 @@ func startAIMediaServer(ls *LivepeerServer) error {
 	ls.HTTPMux.Handle("/audio-to-text", oapiReqValidator(handle(ls, multipartDecoder[worker.GenAudioToTextMultipartRequestBody], processAudioToText)))
 	ls.HTTPMux.Handle("/llm", oapiReqValidator(ls.LLM()))
 	ls.HTTPMux.Handle("/segment-anything-2", oapiReqValidator(handle(ls, multipartDecoder[worker.GenSegmentAnything2MultipartRequestBody], processSegmentAnything2)))
+	ls.HTTPMux.Handle("/realtime-start", ls.RealtimeStart())
```
What will this endpoint be used for? Don't we need just one endpoint like `/realtime-ai` where MediaMTX will send segments? Something similar to what you shared here: https://github.com/livepeer/ai-worker/pull/209#issue-2537592592

j0sh replied:
This is the endpoint that MediaMTX used to signal when a stream starts. Then we can kick off the RTMP pull from there (but that is not in this PR)

leszko commented on server/ai_mediaserver.go:
```diff
@@ -360,3 +361,26 @@ func (ls *LivepeerServer) ImageToVideoResult() http.Handler {
+		req := worker.StartRealtimeToRealtimeFormdataRequestBody{
+			// TODO set model and initial parameters here if necessary (eg, prompt)
+		}
+		resp, err := processAIRequest(ctx, params, req)
```
I don't think we should follow this AI flow here. I guess what we're trying to do is more similar to transcoding.

j0sh replied:
Why not? This selects an orchestrator and allows us to receive and process a response from it (so we know where to send and receive media from)

This method works for now - if you want to change it later as part of the selection work that is fine but this gets us going with the existing RPC mechanisms

mjh1 replied:
I think I see what @leszko is saying, up until now we have just been sending in synchronous AI jobs, the pipeline runs and returns the result. Whereas here the request is to start up the realtime to realtime stream basically? That said I don't think it's a huge issue when there's lots of common code that we're able to reuse by following the same flow.

leszko replied:
Yes, I think we can "re-use" it for this starting of AI runner. But then, the next question would be, when does the runner container stops? Or who stops it? We don't need to solve it in this PR, but something to consider.

j0sh replied:
Yep as @mjh1 says this basically makes a synchronous request using the normal AI inference path. This is pretty convenient, because it gives us an entry point to kick off the rest of the process - see https://github.com/livepeer/go-livepeer/pull/3211 for the beginnings of that.

I expect that we'd be building things like our job monitoring / re-selection, payments, etc somewhere near this '[main loop](https://github.com/livepeer/go-livepeer/pull/3211/files#diff-5dd7b71c3614e76641eba97465018e0d92a2bf6be8918853ac85763e4c2010c6)'

> But then, the next question would be, when does the runner container stops? Or who stops it? We don't need to solve it in this PR, but something to consider.

This is would be a basic RPC request which can be done via whatever call-path makes sense. However, as far as job tear-down goes, I am not sure if we can rely on signaling for that (anything can disappear without notice), so we should ensure things still work well in the absence of explicit tear-down.

We'll need to solve a similar issue with calls to the model control API. The gateway will need some handle to the orchestrator session, the orchestrator to the (remote) worker, and possibly the worker to the runner (if runners have persistent job state). @victorges for visibility since this is his area

leszko replied:
> This is would be a basic RPC request which can be done via whatever call-path makes sense. However, as far as job tear-down goes, I am not sure if we can rely on signaling for that (anything can disappear without notice), so we should ensure things still work well in the absence of explicit tear-down.

Yeah, the best approach from the Distributed (and Decentralized) system perspective is what we do for Transcoding:
1. The params are included in the stream (segments)
2. The AI (Transcoding) pipelines is initialized when the first segment comes in
3. The tear down is down on some timeout (so when there are no segments coming for x min, then it tears down)

Any signaling of warm up / tear down is an "optimization" on top of that. This approach is good for reliability, because a lot may happen, O may go down, we may swap the orchestrator, G may go down (before signaling something), etc. 

CC: @victorges 

Again, I think we can move this discussion outside this PR. The PR can be merged and we can discuss it separately.

leszko reviewed (commented):
Added some comments.

j0sh commented:
Rebased + force pushed to resolve merge conflicts with the latest ai-video. Opening this PR up for real so we can hopefully merge and build on top of it as a team to minimize further conflicts

leszko commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/livepeer/go-livepeer
 
 go 1.21.5
 
+replace github.com/livepeer/ai-worker => github.com/j0sh/livepeer-ai-worker v0.0.0-20241017214425-b4f177ea52a7
```
I think we need to remove it before merging.

j0sh replied:
Yep, actually first we need https://github.com/livepeer/ai-worker/pull/229 or maybe https://github.com/livepeer/ai-worker/pull/233 first in order to define the pipeline schema

leszko commented on server/ai_mediaserver.go:
```diff
@@ -72,6 +72,7 @@ func startAIMediaServer(ls *LivepeerServer) error {
 	ls.HTTPMux.Handle("/audio-to-text", oapiReqValidator(handle(ls, multipartDecoder[worker.GenAudioToTextMultipartRequestBody], processAudioToText)))
 	ls.HTTPMux.Handle("/llm", oapiReqValidator(ls.LLM()))
 	ls.HTTPMux.Handle("/segment-anything-2", oapiReqValidator(handle(ls, multipartDecoder[worker.GenSegmentAnything2MultipartRequestBody], processSegmentAnything2)))
+	ls.HTTPMux.Handle("/live-video-start", ls.StartLiveVideo())
```
```suggestion
         // This is called by Media Server when stream is ready
	ls.HTTPMux.Handle("/live-video-start", ls.StartLiveVideo())
```

1. I'd add the comment that this endpoint is actually called by Media Server
2. Maybe rename to match `/live-video-to-video

j0sh replied:
Comment added in 9e461a7d2ee59b1cd5603382d403c082b71469ad

@victorges [suggested](https://discord.com/channels/423160867534929930/1296851581240610818) using a `/live/` prefix for these paths which sounds good to me so I did that here. Named it `/live/video-to-video/start` since I suspect we will want to append other things to the route, like maybe `/live/video-to-video/api` or thereabouts

leszko commented on server/ai_mediaserver.go:
```diff
@@ -360,3 +361,26 @@ func (ls *LivepeerServer) ImageToVideoResult() http.Handler {
+			// TODO set model and initial parameters here if necessary (eg, prompt)
+		}
+		resp, err := processAIRequest(ctx, params, req)
+		fmt.Println("stream", streamName, "resp", resp, "error", err)
```
If we want to merge it into `ai-video`, then please remove or change to `glog` / `clog`.

j0sh replied:
changed to clog in https://github.com/livepeer/go-livepeer/commit/9e461a7d2ee59b1cd5603382d403c082b71469ad

btw I might send a PR in the near future to add slog-style parameter lists to clog since they are super convenient for ad-hoc KV pairs (in addition to the pre-defined `clog.Add` method we have)

leszko commented on server/ai_http.go:
```diff
@@ -215,6 +216,34 @@ func (h *lphttp) LLM() http.Handler {
+
+		// skipping handleAIRequest for now until we have payments
+
+		// whats the point of openapi if we have to do this manually
```
If you want to merge this PR, then please remove this comment.

j0sh replied:
A bit of late-night frustration that crept through :) These handlers are largely boilerplate that should be done for us by codegen. Removed in https://github.com/livepeer/go-livepeer/commit/9e461a7d2ee59b1cd5603382d403c082b71469ad

leszko commented on server/ai_process.go:
```diff
@@ -864,6 +865,32 @@ func submitAudioToText(ctx context.Context, params aiRequestParams, sess *AISess
+	if resp.JSON500 != nil {
+		fmt.Println("json500", resp.JSON500)
+	}
+	fmt.Println("body is", string(resp.Body))
```
If you plan to merge, then please change to `clog` or remove.

j0sh replied:
Removed in https://github.com/livepeer/go-livepeer/commit/9e461a7d2ee59b1cd5603382d403c082b71469ad

leszko commented on server/ai_process.go:
```diff
@@ -33,6 +33,7 @@ const defaultUpscaleModelID = "stabilityai/stable-diffusion-x4-upscaler"
 const defaultAudioToTextModelID = "openai/whisper-large-v3"
 const defaultLLMModelID = "meta-llama/llama-3.1-8B-Instruct"
 const defaultSegmentAnything2ModelID = "facebook/sam2-hiera-large"
+const defaultLiveVideoToVideoModelID = "stream-diffusion" // TODO what should this be?
```
I guess stream diffusion is not on huggingface, so you can leave it as `stream-diffusion` or change to `cumulo-autumn/stream-diffusion`

j0sh replied:
Changed to `cumulo-autumn/streamdiffusion` in https://github.com/livepeer/go-livepeer/commit/9e461a7d2ee59b1cd5603382d403c082b71469ad

leszko reviewed (approved):
Added some comments. When resolved and CI passes, then I think it's good enough to merge as a scaffold. Thanks @j0sh !

rickstaa commented:
Hey @j0sh, just a heads up—I had to update the dependencies to ensure the network is compatible with the new Go version and other updated libraries. I found a combination that works, but if you need to use lower versions that are also compatible, feel free to adjust them. I'm flexible with it. I'm also improving the code through small pull requests so that it doesn't cause to many conflicts. If you run into problems or need me to rebase just ping me.

j0sh commented:
Force pushed to fix up a bunch of merge conflicts with the latest ai-video branch

Also squashed everything down to 1 commit to avoid a cascading series of merge conflicts

Builds for this PR are broken for now until the pipeline on the worker / runner side is in, either in https://github.com/livepeer/ai-worker/pull/229 or https://github.com/livepeer/ai-worker/pull/233

j0sh committed:
Remove dependency on ai-worker changes

j0sh commented:
Updated to remove the dependency on AI Worker changes. The scaffolding that is there now is now even less useful, but there is still the capability definition and pipeline path, both of which have caused merge conflicts recently as we've added new stuff

If CI is green here, I'll go ahead and merge so it's easier to stay on the leading edge of the tree.
